### ISSUE TITLE
lazy() via cursor fetching

### ISSUE THREAD

**Issue by leio** [tpetry/laravel-postgresql-enhanced#11]
Eloquents `lazy()` has the problem of making new queries for each chunk via OFFSET+LIMIT, which doesn't give a stable result if something changed in parallel. Meanwhile `lazyById` prevents applying own ordering, plus aren't completely safe from changes in parallel requests either. And `chunk()` ends up with all the results queried out into memory at PDO level, which is unsuitable for dealing with a lot of rows.

So I had to come up with a version that overcomes all that by making use of PostgreSQL cursors to handle the chunking.
Here's a quick (and perhaps a bit dirty in regards to toSql() usage) macro that does this:

```php
    public function boot()
    {
        Builder::macro('lazyFetch', function(int $chunkSize = 1000) {
            if ($chunkSize < 1) {
                throw new InvalidArgumentException('The chunk size should be at least 1');
            }
            if (DB::transactionLevel() <= 0) {
                throw new RuntimeException('lazyFetch can only be used within a DB transaction');
            }

            return LazyCollection::make(function () use ($chunkSize) {
                $cursorName = 'lazy_fetch_' . spl_object_id($this);
                DB::statement("DECLARE {$cursorName} NO SCROLL CURSOR FOR " . $this->toSql(), $this->getBindings());

                while (true) {
                    $results = DB::select("FETCH FORWARD {$chunkSize} FROM {$cursorName}");

                    foreach ($results as $result) {
                        yield $result;
                    }

                    if (count($results) < $chunkSize) {
                        DB::statement("CLOSE {$cursorName}");
                        return;
                    }
                }
            });
        });
    }
```

and then it can be used by just using `lazyFetch()` instead of `lazy()` to get a stable LazyCollection. It might even be a good idea to use that for `lazy()` automatically if within a transaction?

Anyhow, I think it might be a good fit into laravel-postgresql-enhanced in some form. If so, how would it be integrated into a package like laravel-postgresql-enhanced is for easy usage by programs?
Might take a stab at making a PR for this if I knew which way to integrate it into the package.

**Comment by tpetry** [comment]
So basically you are suggesting a `lazyByCursor` function? That‘s a good idea. I could work on it based on your draft. Just needs a bit of polishing. But transactions are not really needed or did i miss a detail?

**Comment by leio** [comment]
PostgreSQL will give you an unceremonial

> ERROR:  DECLARE CURSOR can only be used in transaction blocks

if you aren't in a transaction, which makes sense I suppose.

The main thing in my macro-based approach I'm not really sure about is:
* what happens if the generator gets deleted without completion. The cursor could stick around and the next lazyByCursor call could end up with the same `spl_object_id` and give an error (this is something I saw in completing cases too until I added the `CLOSE` call). Shouldn't be a problem across different PHP processes, as they'd have a separate DB connection, and the cursor names appear to be OK to be the same across separate DB connections per quick psql tests.

**Comment by tpetry** [comment]
I looked in the Laravel source code again to be sure. Can‘t you just use `cursor()`? I have used it before to lazy load a lot of rows without getting memory exceed errors.

**Comment by leio** [comment]
It doesn't have any control on how much it fetched from SQL at a time, so it appears to fetch it all and I had unbound memory usage going on for a huge dataset I was processing in a streaming manner. With this lazyFetch I was in control of the memory usage bound (how big chunk value I use vs how many rows I get per roundtrip to DB) and I could keep to a maximum memory usage. cursor() bounds the memory usage in the processed form (ORM overhead), but doesn't seem to help in controlling the raw datas memory use.

**Comment by tpetry** [comment]
I have now been playing around an hour with the `cursor()` method and I can say that it is definitely not loading into memory. I am using it in my applications also to circumvent memory limits. 

Quick&Dirty example (dataset with 945MB):
```php
ini_set('memory_limit', '32M');

class Example extends Model {
    public $table = 'example';
    protected $guarded = [];
    public $timestamps = false;
};

DB::unprepared('
  CREATE TABLE IF NOT EXISTS example (
    id bigint NOT NULL GENERATED ALWAYS AS IDENTITY PRIMARY KEY,
    str text NOT NULL
  );
');

dump('Seeding table...');
DB::unprepared('INSERT INTO example (str) SELECT md5(random()::text) FROM generate_series(1, 1000000)');

dump('Iterating table...');
foreach (Example::cursor() as $i => $example) {
  if ($i % 1000 === 0) {
    dump($i);
    dump([memory_get_usage(true), memory_get_peak_usage(true)]);
  }
}
```

So if you do get memory errors it's a problem in your application code.

But I see benefits of your approach:
* The `cursor()` method only starts looping when the query is finished? So it's not loading everything into memory, but still the first item is only looped after a very long time - with the cursor approach it's instant.
* Cursors have the additional benefit of a special language construct for `UPDATE/DELETE` statements you can execute a DML statement on them efficiently  with`WHERE CURRENT OF cursor` because the db already knows the location.

I will look further into the cursor feature, thanks for suggesting it! In the next days/weeks I have to research what's happening with the cursor and it's pagination when you modify the table in the same transaction you are currently using for looping. If there's no problem I will add it, if there are some problems I will have to further look at the feature.

**Comment by leio** [comment]
I am very sure that there are memory issues with `cursor()` and not my application, I worked quite a bit on measurements and benchmarking things for my use case. If I load a huge dataset with cursor and put in a sleep before the last item is iterated, I could see 300MB+ heap RAM usage by PHP - if I do the same with my lazyFetch(), it peaked out at 80-100MB with my dataset (60MB being base usage by the PHP process without doing any of it), because it is memory-equal to running the same query with `cursor()` that has an additional `limit()` applied to it that is equal to the `$chunkSize` in case of `lazyFetch`. If I used smaller dataset with cursor(), the peak memory usage was reduced; if larger, it grew — with lazyFetch, it was under my own control, depending on what my `$chunkSize` is.

Don't look at the memory usage with `memory_get_usage` or `memory_get_peak_usage` — these won't show this, it's memory usage further down the stack, which these PHP functions do not capture, because the memory in question here probably isn't allocated with `emalloc` (maybe it's even in libpq if not PDO). Use operating system level measuring of the memory usage to see this.

There is some talks about it in e.g. https://stackoverflow.com/questions/45464676/what-is-the-difference-between-laravel-cursor-and-laravel-chunk-method#comment101130039_53736557 that seems relevant and matching to what I see.

**Comment by tpetry** [comment]
As I've could have used that exact feature today (`lazyByCursor` with `whereCursor`) I will definitely implement it! I don't have time for it this week but will do it in the next weeks. I need some time to play around on how to clean up the cursor if the user inserts a `break` to skip the remaining the cursor results etc.

@leio Thanks already for that great idea, it will be a perfect fit!

**Comment by leio** [comment]
I guess `lazyByCursor` + `whereCursor` will be a bit tricky, because the cursor would be at `$chunkSize` offsets, not exactly where you are with the generators consumption/iteration.

Could work when `$chunkSize` is 1, but then it's a lot of roundtrips with the server (if it's a dedicated separate DB server).

Maybe, with some extra bookkeeping, `whereCursor` usage could trigger a cursor back movement (then changed to not be a `NO SCROLL` one)  to where the generator yielding is at, while the next cursor forward fetch in the `lazyByCursor` generator loop would be preceded by moving it back forward to the right position based on how much `whereCursor` had pushed it forward?

Not sure what the use cases are — where the benefit of it is over just doing things based on the row in question and its pkey.

**Comment by tpetry** [comment]
Thanks for your feature idea @leio, it's implemented in release `0.13.0`. I found a way to do a more clean management of cursors and found a trick to clean up the stuff. If you put all the code in a `try ... finally` statement the finally part will be called when `break`ing the generator. So the cursor is always cleaned up correctly :)
